### ISSUE TITLE
Custom layer with tf.extract_image_patches extremely slow

### ISSUE THREAD

**Issue by Leonamrsm** [tensorflow/tensorflow#58086]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Performance

### Source

binary

### Tensorflow Version

2.9.2

### Custom Code

Yes

### OS Platform and Distribution

Google Colab

### Python version

3.7.14

### Current Behaviour?

```shell
I'm new to tensorflow. I'm trying to implement a custom pooling layer, using owa operators (https://github.com/jiforcen/ordered-weighted-pooling). For that I'm using tf.extract_image_patches, but this operation is extremely slow when the input data dimensions are large, as raised in issue #13017. 

I believe that the pooling layer that I implemented has a behavior very similar to what is performed by the tf.keras.layers.MaxPooling2D layer. Inspecting the code of MaxPooling2D I saw that it calls the gen_nn_ops.max_pool method. 

I tried to take a look at what's inside gen_nn_ops.max_pool, but I can't find it in the repository. From what I've googled I can't find the source code because it's automatically generated by bazel. If I build from source, I'll see this file inside bazel-genfiles, right? This file contains automatically generated Python wrappers to underlying C++ implementations.

Is it possible to create a custom pooling operation in C++ and use it in my Python code? I'm adding the custom layer code that I implemented.
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

from keras import backend as K
from tensorflow.keras.constraints import Constraint
from tensorflow.python.util.tf_export import keras_export
from tensorflow.python.keras.utils import conv_utils
# import skimage.measure

@keras_export('keras.constraints.UnitSumNonNeg', 'keras.constraints.unit_sum_non_neg')
class UnitSumNonNeg(Constraint):
    """Limits weights to be non-negative and with sum equal to one

    Also available via the shortcut function `keras.constraints.unit_sum_non_neg`.
    """
    def __call__(self, w):
        aux =  w * tf.cast(tf.math.greater_equal(w, 0.), w.dtype)

        return aux/(K.epsilon() + tf.reduce_sum(aux, axis=[0], keepdims=True))

class OWAPoolingNew(tf.keras.layers.Layer):
    def __init__(self,
               pool_size=(2, 2),
               strides=None,
               padding='valid',
               data_format=None,
               name=None,
               sort=True,
               train=True, 
               seed=None,
               all_channels=False,
               **kwargs):
        super(OWAPoolingNew, self).__init__(name=name, **kwargs)

        self.pool_size = pool_size
        self.strides = pool_size if strides == None else strides
        self.padding = padding
        self.data_format = conv_utils.normalize_data_format('channels_last')
        self.sort = sort
        self.train = train
        self.seed = seed if seed != None else 10
        self.all_channels = all_channels
        
    def build(self, input_shape):
      
      if self.all_channels:
        weights_shape = (self.pool_size[0] * self.pool_size[1], input.shape[-1])
      else:
        weights_shape = (self.pool_size[0] * self.pool_size[1], 1)
      
      tf.random.set_seed(self.seed)
      kernel = tf.random.uniform(shape=weights_shape)
      kernel /= tf.reduce_sum(kernel, axis=[0], keepdims=True)
      
      self.kernel = tf.Variable(initial_value = kernel, trainable=self.train, dtype='float32', constraint=UnitSumNonNeg())

    def call(self, inputs):

        _, height, width, channels = inputs.get_shape().as_list()

        # Extract pooling regions
        stride = [1, self.strides[0], self.strides[1], 1]
        ksize = [1, self.pool_size[0], self.pool_size[1], 1]

        inputs = tf.image.extract_patches(inputs, sizes = ksize, strides = stride,
                            rates = [1, 1, 1, 1], padding='SAME')

        _, pool_height, pool_width, elems = inputs.get_shape().as_list()

        # Extract pooling regions for each channel
        elems =  int(elems / channels)
        inputs = tf.reshape(inputs, [-1, pool_height, pool_width, elems, channels]) # Reshape tensor

        # Sort values for pooling
        if self.sort:
            inputs = tf.sort(inputs, axis=-2, direction='DESCENDING', name=None)

        outputs = tf.reduce_sum(tf.math.multiply(self.kernel, inputs), axis=-2)

        return outputs
```

### Relevant log output

_No response_</details>

**Comment by tilakrayal** [comment]
@Leonamrsm,
I tried to execute the code on tensorflow v2.9.2 and haven't found any perfomance issue during the execution. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/140b18ae26a904d57271f9846202507e/untitled686.ipynb) and let us know if anything is missing here. Thank you!

**Comment by Leonamrsm** [comment]
@tilakrayal 
Backprop through extract_image_patches is extremly slow, specifically the transpose call. It appears to be this specific call: tensorflow/python/ops/array_grad.py#L747

![image](https://user-images.githubusercontent.com/52931821/195832260-e247c50c-ac5a-449e-b53d-52b3527bd058.png)

I updated the [gist ](https://gist.github.com/Leonamrsm/d07faff69e5d5097f54bd4149e796378#file-untitled686-ipynb) with a call from the new layer and the MaxPooling2D layer from keras, using the training data from the cifar10 dataset with dimensions 5000x32x32x3. When using the new layer it took 42s to execute and when using MaxPooling2D 2s.

**Comment by tilakrayal** [comment]
@sachinprasadhs,
I was able to reproduce the issue on tensorflow v2.9, v2.10 and [nightly](https://colab.research.google.com/gist/tilakrayal/3455a7ccb2fbd05d7d73ffcdddb61899/untitled686.ipynb). Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/c086fdee982e02e6e8d2cc67a8cfc889/untitled686.ipynb).

**Comment by cantonios** [comment]
> I tried to take a look at what's inside gen_nn_ops.max_pool, but I can't find it in the repository.

All these `gen_**_ops` get automatically generated from the op definitions, e.g. [nn_ops.cc:870](https://github.com/tensorflow/tensorflow/blob/f56dd27d3e94d15af6ca82e3e5a5c4fbaf34771d/tensorflow/core/ops/nn_ops.cc#L870) for "MaxPool".  The generated python code won't help you much - it just creates a wrapper around a call to the C++ kernel.  The actual kernel that gets called depends on the device you're running it on.  For CPU,
- The `MaxPoolingOp` kernel is registered [here](https://github.com/tensorflow/tensorflow/blob/f56dd27d3e94d15af6ca82e3e5a5c4fbaf34771d/tensorflow/core/kernels/maxpooling_op.cc#L1551).
- The definition of `MaxPoolingOp` is [here](https://github.com/tensorflow/tensorflow/blob/f56dd27d3e94d15af6ca82e3e5a5c4fbaf34771d/tensorflow/core/kernels/pooling_ops_common.h#L90).

These ops also have special gradients defined.  For "MaxPool", the gradient definition is [here](https://github.com/tensorflow/tensorflow/blob/f56dd27d3e94d15af6ca82e3e5a5c4fbaf34771d/tensorflow/python/ops/nn_grad.py#L676), which just calls the wrapped gradient op "MaxPoolGrad", which has the kernel [`MaxPoolingGradOp`](https://github.com/tensorflow/tensorflow/blob/f56dd27d3e94d15af6ca82e3e5a5c4fbaf34771d/tensorflow/core/kernels/maxpooling_op.cc#L225).  

> Is it possible to create a custom pooling operation in C++ and use it in my Python code?

Yes, it would be similar to the above.  See the custom op [guide](https://www.tensorflow.org/guide/create_op).  The gradient can also be a custom op, just as it is for `MaxPool`.

**Comment by Leonamrsm** [comment]
@cantonios 
Thanks for the answer. I believe this solution should work for me. I will follow the [guide ](https://www.tensorflow.org/guide/create_op )to create a custom operation.

**Comment by cantonios** [comment]
Marking as closed then.  For the slow transpose issue, if it's a blocker, please re-open #13017.
